lib/Dialect/TritonNvidiaGPU/Transforms/MaterializeLoadStore.cpp

```diff
@@ -202,6 +219,90 @@ void MaterializeLoadStorePass::materializeStoreTilePtr(
   store->erase();
 }
 
+SmallVector<Value>
+MaterializeLoadStorePass::immutableMbarrierAlloc(Operation *entryOp,
+                                                 unsigned loadNum) {
+  SmallVector<Value> barrierLists;
+  OpBuilder builder(entryOp);
+  Location loc = entryOp->getLoc();
+  builder.setInsertionPoint(entryOp);
+  for (unsigned i = 0; i < loadNum; ++i) {
+    auto mBarrierTy =
+        mlir::triton::PointerType::get(builder.getIntegerType(64), 3);
+    Value mBarrier = builder.create<ttng::AllocMBarrierOp>(loc, mBarrierTy, 1);
+    barrierLists.push_back(mBarrier);
+  }
+  return barrierLists;
+}
+
+/// Func Brief
+///   getPhase function inserts phase to every nested forOp recursively and
+///   returns the phase loop-carried by current forOp.
+/// Design Specification
+///   Each loadOp phase need to be loop-carried by all outside forOps.
+///   Just like this structure:
+///   scf.for ... iter_args(%arg2, %arg1)
+///     loadOp
+///     mbarrier_wait(%arg1)
+///     scf.for ... iter_args(%arg2)
+///       loadOp
+///       mbarrier_wait(%arg2)
+///       %arg3 = %arg2 ^ 1
+///       yield(%arg3)
+///     %arg4 = %arg1 ^ 1
+///     yiled(%arg3, %arg4)
+///   In this structure, inside forOp operand depends on outside forOp
+///   blockArgument, so the forOp recreate order is from outside to inside.
+///   BTW, the yieldOp of outside forOp depends on the result of inside forOp,
+///   so the yielOp insertOperand will delay one iteration.
+Value MaterializeLoadStorePass::getPhase(scf::ForOp forOp, bool initCall) {
+  OpBuilder builder(forOp);
+  Location loc = forOp.getLoc();
+
+  // Step 1: get initVal
+  Value initVal;
+  if (auto parentForOp = llvm::dyn_cast<scf::ForOp>(forOp->getParentOp()))
```

Comment:
getParentOp() should be replaced with getParentOfType<ForOp>().

For example:
```
   scf.for () {
      scf.if () {
          scf.for () {
            ...
          }
      }
   }
```

Reply:
All right, modified.